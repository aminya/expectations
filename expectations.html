<!DOCTYPE html>
<html>
<head>
        <title>expectations (expectations)</title>
	<meta charset="utf-8" />
	<meta content="width=device-width, initial-scale=1" name="viewport" />
        <link rel="stylesheet" href="style.css" />
        <script type="text/javascript" src="script.js"></script>

	
	<link rel="prefetch" href="search-results.html" />
</head>
<body>
	<div id="page-header">
		<div id="logotype">
		<span>Documentation</span>
		<nav>
			<a href="http://dlang.org/">Dlang.org</a>
		</nav>
		</div>

		<form action="search-docs.html" id="search">
			<input type="search" placeholder="Find a symbol name..." name="searchTerm" />
			<input type="submit" value="Go" />
		</form>
	</div>
	<div id="page-body">
		<div id="page-content">
		<h1>expectations</h1><div class="breadcrumbs"></div><div><div class="documentation-comment synopsis"><div><p>Error handling that bundles exceptions with return values.</p><p>The design of this module is based on C++'s proposed
<a href="https://wg21.link/p0323">std::expected</a> and Rust's
<a href="https://doc.rust-lang.org/std/result/">std::result</a>. See
<a href="https://www.youtube.com/watch?v=nVzgkepAg5Y">&quot;Expect the Expected&quot;</a> by
Andrei Alexandrescu for further background.</p></div></div></div><div class="annotated-prototype"></div><h2 id="members"><a href="#members" class="header-anchor">Members</a></h2><h3 id="class" class="member-list-header hide-from-toc"><a href="#class" class="header-anchor">Classes</a></h3><dl class="member-list native"><dt><a href="expectations.Unexpected.html">Unexpected</a><div style="max-width: 20ch;" class="simplified-prototype"><span class="builtin-type">class</span> <span class="name">Unexpected</span><span class="template-params">(T)</span></div></dt><dd><div><p>An exception that represents an error value.</p></div></dd></dl><h3 id="function" class="member-list-header hide-from-toc"><a href="#function" class="header-anchor">Functions</a></h3><dl class="member-list native"><dt><a href="expectations.expected.html">expected</a><div style="max-width: 36ch;" class="simplified-prototype"><tt class="highlighted"><span class="hid">Expected</span>!(<span class="hid">T</span>, <span class="hid">E</span>)</tt> <span class="name">expected</span><tt class="highlighted">(<span class="hid">T</span> <span class="hid">value</span>)</tt></div></dt><dd><div><p>Creates an <tt class="inline-code">Expected</tt> object from an expected value, with type inference.</p></div></dd><dt><a href="expectations.missing.html">missing</a><div style="max-width: 33ch;" class="simplified-prototype"><tt class="highlighted"><span class="hid">Expected</span>!(<span class="hid">T</span>, <span class="hid">E</span>)</tt> <span class="name">missing</span><tt class="highlighted">(<span class="hid">E</span> <span class="hid">err</span>)</tt></div></dt><dd><div><p>Creates an <tt class="inline-code">Expected</tt> object from an error value.</p></div></dd></dl><h3 id="struct" class="member-list-header hide-from-toc"><a href="#struct" class="header-anchor">Structs</a></h3><dl class="member-list native"><dt><a href="expectations.Expected.html">Expected</a><div style="max-width: 36ch;" class="simplified-prototype"><span class="builtin-type">struct</span> <span class="name">Expected</span><span class="template-params">(T, E = Exception)</span></div></dt><dd><div><p>An <tt class="inline-code">Expected!(T, E)</tt> contains either an expected value of type <tt class="inline-code">T</tt>, or an
error value of type <tt class="inline-code">E</tt> explaining why the expected value couldn't be
produced.</p></div></dd></dl><h3 id="template" class="member-list-header hide-from-toc"><a href="#template" class="header-anchor">Templates</a></h3><dl class="member-list native"><dt><a href="expectations.flatMap.html">flatMap</a><div style="max-width: 29ch;" class="simplified-prototype"><span class="builtin-type">template</span> <span class="name">flatMap</span><span class="template-params">(alias fun)</span></div></dt><dd><div><p>Forwards the expected value in an <tt class="inline-code">Expected</tt> object to a function that
returns an <tt class="inline-code">Expected</tt> result.</p></div></dd><dt><a href="expectations.map.html">map</a><div style="max-width: 25ch;" class="simplified-prototype"><span class="builtin-type">template</span> <span class="name">map</span><span class="template-params">(alias fun)</span></div></dt><dd><div><p>Applies a function to the expected value in an <tt class="inline-code">Expected</tt> object.</p></div></dd></dl><div><h2 id="examples"><a href="#examples" class="header-anchor">Examples</a></h2><div class="documentation-comment"><div></div></div><div class="unittest-example-holder"><div class="documentation-comment"><h3 id="basic-usage" class="user-header"><a href="#basic-usage" class="header-anchor">Basic Usage</a></h3></div><pre class="d_code highlighted with-line-wrappers"><span class="br">1 </span><span class="kwrd">import</span> <span class="hid">std</span>.<span class="hid">exception</span>: <span class="hid">assertThrown</span>;
<span class="br">2 </span>
<span class="br">3 </span><span class="hid">Expected</span>!<span class="type">int</span> <span class="hid">charToDigit</span>(<span class="type">char</span> <span class="hid">c</span>)
<span class="br">4 </span>{
<span class="br">5 </span>    <span class="type">int</span> <span class="hid">d</span> = <span class="hid">c</span> - <span class="str">'0'</span>;
<span class="br">6 </span>    <span class="kwrd">if</span> (<span class="hid">d</span> &gt;= <span class="num">0</span> &amp;&amp; <span class="hid">d</span> &lt; <span class="num">10</span>) {
<span class="br">7 </span>        <span class="kwrd">return</span> <span class="hid">expected</span>(<span class="hid">d</span>);
<span class="br">8 </span>    } <span class="kwrd">else</span> {
<span class="br">9 </span>        <span class="kwrd">return</span> <span class="hid">missing</span>!<span class="type">int</span>(
<span class="br">10 </span>            <span class="kwrd">new</span> <span class="hid">Exception</span>(<span class="hid">c</span> ~ <span class="str">&quot; is not a valid digit&quot;</span>)
<span class="br">11 </span>        );
<span class="br">12 </span>    }
<span class="br">13 </span>}
<span class="br">14 </span>
<span class="br">15 </span><span class="kwrd">auto</span> <span class="hid">goodResult</span> = <span class="hid">charToDigit</span>(<span class="str">'7'</span>);
<span class="br">16 </span><span class="kwrd">auto</span> <span class="hid">badResult</span> = <span class="hid">charToDigit</span>(<span class="str">'&amp;'</span>);
<span class="br">17 </span>
<span class="br">18 </span><span class="kwrd">assert</span>(<span class="hid">goodResult</span>.<span class="hid">hasValue</span>);
<span class="br">19 </span><span class="kwrd">assert</span>(<span class="hid">goodResult</span>.<span class="hid">value</span> == <span class="num">7</span>);
<span class="br">20 </span>
<span class="br">21 </span><span class="kwrd">assert</span>(!<span class="hid">badResult</span>.<span class="hid">hasValue</span>);
<span class="br">22 </span><span class="hid">assertThrown</span>(<span class="hid">badResult</span>.<span class="hid">value</span>);
<span class="br">23 </span><span class="kwrd">assert</span>(<span class="hid">badResult</span>.<span class="hid">error</span>.<span class="hid">msg</span> == <span class="str">&quot;&amp; is not a valid digit&quot;</span>);</pre></div><h2 id="meta"><a href="#meta" class="header-anchor">Meta</a></h2><div class="documentation-comment source-section other-section"><h3 id="source"><a href="#source" class="header-anchor">Source</a></h3><div><p><a href="source/expectations.d.html">See Source File</a><br /></p></div></div><div class="documentation-comment license-section other-section"><h3 id="license"><a href="#license" class="header-anchor">License</a></h3><div><p>MIT</p></div></div><div class="documentation-comment authors-section other-section"><h3 id="authors"><a href="#authors" class="header-anchor">Authors</a></h3><div><p>Paul Backus</p></div></div></div></div>
		<div id="page-nav">
		<span class="type-separator">modules</span><ul><li><a href="expectations.html" class="module current">expectations</a></li></ul></div>
	</div>
	<div id="page-footer">Page generated by <a href="https://github.com/adamdruppe/adrdox">adrdox</a></div>
</body>
</html>