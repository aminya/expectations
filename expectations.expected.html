<!DOCTYPE html>
<html>
<head>
        <title>expected (expectations.expected)</title>
	<meta charset="utf-8" />
	<meta content="width=device-width, initial-scale=1" name="viewport" />
        <link rel="stylesheet" href="style.css" />
        <script type="text/javascript" src="script.js"></script>

	
	<link rel="prefetch" href="search-results.html" />
</head>
<body>
	<div id="page-header">
		<div id="logotype">
		<span>Documentation</span>
		<nav>
			<a href="http://dlang.org/">Dlang.org</a>
		</nav>
		</div>

		<form action="search-docs.html" id="search">
			<input type="search" placeholder="Find a symbol name..." name="searchTerm" />
			<input type="submit" value="Go" />
		</form>
	</div>
	<div id="page-body">
		<div id="page-content">
		<h1>expected</h1><div class="breadcrumbs"><a href="expectations.html" class="breadcrumb">expectations</a></div><div><div class="documentation-comment synopsis"><div><p>Creates an <tt class="inline-code">Expected</tt> object from an expected value, with type inference.</p></div></div></div><div class="annotated-prototype"><div class="function-prototype"><div class="attributes"></div><div class="return-type"><a data-ident="Expected" href="expectations.Expected.html" title="expectations.Expected" class="name">Expected</a>!(<span data-ident="T" class="name">T</span>, <span data-ident="E" class="name">E</span>)</div><div class="function-name">expected</div><div class="template-parameters">(<div class="parameters-list toplevel"><div class="template-parameter-item parameter-item">	<span><span data-ident="T" class="name">T</span></span></div><div class="template-parameter-item parameter-item">	<span><span data-ident="E" class="name">E</span> = <span data-ident="Exception" class="name">Exception</span></span></div></div>)</div><div class="runtime-parameters">(<div class="parameters-list toplevel"><div class="runtime-parameter-item parameter-item"><span class="parameter-type-holder"><span class="parameter-type"><span data-ident="T" class="name">T</span></span></span> <span data-ident="value" class="parameter-name name"><a href="#param-value">value</a></span></div></div>)</div><div class="function-contracts"></div></div></div><div><h2 id="meta"><a href="#meta" class="header-anchor">Meta</a></h2><div class="documentation-comment source-section other-section"><h3 id="source"><a href="#source" class="header-anchor">Source</a></h3><div><p><a href="source/expectations.d.html#L367">See Implementation</a><br /></p></div></div></div></div>
		<div id="page-nav"><a href="expectations.html" class="parent">expectations</a>
		<span class="type-separator">classes</span><ul><li><a href="expectations.Unexpected.html" class="class">Unexpected</a></li></ul><span class="type-separator">functions</span><ul><li><a href="expectations.expected.html" class="function current">expected</a></li><li><a href="expectations.missing.html" class="function">missing</a></li></ul><span class="type-separator">structs</span><ul><li><a href="expectations.Expected.html" class="struct">Expected</a></li></ul><span class="type-separator">templates</span><ul><li><a href="expectations.flatMap.html" class="template">flatMap</a></li><li><a href="expectations.map.html" class="template">map</a></li></ul></div>
	</div>
	<div id="page-footer">Page generated by <a href="https://github.com/adamdruppe/adrdox">adrdox</a></div>
</body>
</html>